### PR TITLE
Add feature: can use type ReactNode for message of rules (#10136)

### PR DESCRIPTION
**Pull request by lovekonakona**
First of all, thank you for your contribution! :-)

Please makes sure that these checkboxes are checked before submitting your PR, thank you!

* [x] Make sure that you propose PR to right branch: bugfix for `master`, feature for latest active branch `feature-x.x`.
* [x] Make sure that you follow antd's [code convention](https://github.com/ant-design/ant-design/wiki/Code-convention-for-antd).
* [x] Run `npm run lint` and fix those errors before submitting in order to keep consistent code style.
* [ ] Rebase before creating a PR to keep commit history clear.
* [x] Add some descriptions and refer relative issues for you PR.

Extra checklist:

**if** *isBugFix* **:**

  * [ ] Make sure that you add at least one unit test for the bug which you had fixed.

**elif** *isNewFeature* **:**

  * [x] Update API docs for the component.
  * [ ] Update/Add demo to demonstrate new feature.
  * [ ] Update TypeScript definition for the component.
  * [x] Add unit tests for the feature.

### CODE REVIEW

File: components/form/FormItem.tsx [hunk 1]
```diff
@@ -70,7 +70,16 @@ export default class FormItem extends React.Component<FormItemProps, any> {
     const onlyControl = this.getOnlyControl();
     if (props.help === undefined && onlyControl) {
       const errors = this.getField().errors;
-      return errors ? errors.map((e: any) => e.message).join(', ') : '';
+      if (errors) {
+        return errors.map((e: any, index: number) => {
+          return <React.Fragment key={index}>
+            {index > 0 && ', '}
+            {e.message}
+          </React.Fragment>;
```
Comment (yesmeck) [inline components/form/FormItem.tsx]: I don't think we can use `React.Fragment`

**Comment by afc163** [comment]
ref #10136

File: components/form/FormItem.tsx (hunk 1 above)
Comment (lovekonakona) [reply components/form/FormItem.tsx]: I can use `span` here?

File: components/form/FormItem.tsx [hunk 2]
```diff
@@ -70,7 +70,16 @@ export default class FormItem extends React.Component<FormItemProps, any> {
     const onlyControl = this.getOnlyControl();
     if (props.help === undefined && onlyControl) {
       const errors = this.getField().errors;
-      return errors ? errors.map((e: any) => e.message).join(', ') : '';
+      if (errors) {
+        return errors.map((e: any, index: number) => {
+          return <span key={index}>
+            {index > 0 && ', '}
+            {e.message}
+          </span>;
```
Comment (yesmeck) [inline components/form/FormItem.tsx]: How about `React.isElement(e.message) ? React.cloneElement(e.message, { key: index }) : e.message`?

File: components/form/demo/custom-message.md [hunk 3]
```diff
@@ -0,0 +1,65 @@
+---
```
Comment (yesmeck) [inline components/form/demo/custom-message.md]: 不是很复杂的使用，感觉没必要单独写个 demo?

File: components/form/__tests__/message.test.js [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+    const wrapper = mount(<Form1 />);
+    myForm.validateFields();
+
+    wrapper.update();
+
+    expect(wrapper.find('.ant-form-explain').length).toBe(1);
+    expect(wrapper.find('.ant-form-explain').at(0).getDOMNode().textContent.trim()).toBe('Account does not exist, Forgot account?');
+    const link = wrapper.find('.ant-form-explain a').at(0).getDOMNode();
+    expect(link.textContent.trim()).toBe('Forgot account?');
+    expect(link.href).toBe('https://www.alipay.com/');
```
Comment (yesmeck) [inline components/form/__tests__/message.test.js]: 其实比较简单的做法是 `expect(wrapper).toMatchSnapshot()` 就可以了。

Comment (yesmeck) [reply components/form/__tests__/message.test.js]: 不需要写那么多判断

**Comment by yesmeck** [comment]
👍

File: components/form/__tests__/message.test.js [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+          </Form.Item>
+        </Form>
+      );
+    });
+
+    const wrapper = mount(<Form1 />);
+    myForm.validateFields();
+
+    wrapper.update();
+    expect(wrapper).toMatchSnapshot();
```
Comment (yesmeck) [inline components/form/__tests__/message.test.js]: Last thing `expect(wrapper.render()).toMatchSnapshot()` 😄

File: components/form/FormItem.tsx [hunk 6]
```diff
@@ -70,7 +70,22 @@ export default class FormItem extends React.Component<FormItemProps, any> {
...
+          if (React.isValidElement(e.message)) {
+            messages.push(React.cloneElement(e.message, { key: index }));
+          } else {
+            messages.push(e.message);
+          }
+          if (index < errors.length - 1) {
+            messages.push(', ');
+          }
+        });
+        return messages;
```
Comment (afc163) [inline components/form/FormItem.tsx]: Try `map` over `forEach` and `push`

**Review by yesmeck — APPROVED** [review]

File: components/form/FormItem.tsx (hunk 6 above)
Comment (yesmeck) [reply components/form/FormItem.tsx]: `reduce`?

File: components/form/__tests__/__snapshots__/demo.test.js.snap [hunk 7]
```diff
@@ -46,6 +46,9 @@ exports[`renders ./components/form/demo/advanced-search.md correctly 1`] = `
                   value=""
                 />
               </span>
+              <div
+                class="ant-form-explain"
+              />
```
Comment (afc163) [inline components/form/__tests__/__snapshots__/demo.test.js.snap]: errors 那里的判断貌似还是要加上，否则这里会打出不必要的空 div。

File: components/form/FormItem.tsx [hunk 8]
```diff
@@ -2,6 +2,7 @@ import * as React from 'react';
 import * as ReactDOM from 'react-dom';
 import PropTypes from 'prop-types';
 import classNames from 'classnames';
+import intersperse from 'intersperse';
```
Comment (yesmeck) [inline components/form/FormItem.tsx]: 🤔

File: components/form/FormItem.tsx [hunk 9]
```diff
@@ -65,15 +66,20 @@ export default class FormItem extends React.Component<FormItemProps, any> {
...
+    const { help } = this.props;
+    if (help === undefined && this.getOnlyControl()) {
       const errors = this.getField().errors;
-      return errors ? errors.map((e: any) => e.message).join(', ') : '';
+      if (errors) {
+        return intersperse(errors.map((e: any, index: number) => (
+          React.isValidElement(e.message)
+            ? React.cloneElement(e.message, { key: index })
+            : e.message
+        )), ' ');
```
Comment (yesmeck) [inline components/form/FormItem.tsx]: 逗号

**Pull request merged**